presto-main/src/main/java/com/facebook/presto/execution/buffer/AESSpillCipher.java

```diff
@@ -0,0 +1,219 @@
+
+import java.nio.ByteBuffer;
+import java.security.GeneralSecurityException;
+import java.security.NoSuchAlgorithmException;
+
+import static com.facebook.presto.spi.StandardErrorCode.GENERIC_INTERNAL_ERROR;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+final class AESSpillCipher
+        implements SpillCipher
+{
+    public enum Mode
+    {
+        AES_128(128),
+        AES_256(256);
+
+        private final int keyBits;
+
+        Mode(int keyBits)
+        {
+            this.keyBits = keyBits;
+        }
+
+        public int getKeyBits()
+        {
+            return keyBits;
+        }
+    }
+
+    private SecretKeySpec key;
+    private final int keyBits;
+    private final int ivBytes;
+
+    AESSpillCipher(Mode mode)
+    {
+        this.keyBits = requireNonNull(mode, "mode is null").getKeyBits();
+        this.key = generateNewSecretKey(keyBits);
+        this.ivBytes = createEncryptCipher().getIV().length;
+    }
+
+    @Override
+    public ByteBuffer decrypt(ByteBuffer cipherText)
+    {
+        int length = cipherText.remaining();
+        if (length < ivBytes) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, format("cipherText too small to contain IV. Minimum: %s, Actual: %s", ivBytes, length));
+        }
+        byte[] iv = new byte[ivBytes];
+        cipherText.get(iv);
+        Cipher cipher = createDecryptCipher(new IvParameterSpec(iv));
+        ByteBuffer output = ByteBuffer.allocate(cipher.getOutputSize(cipherText.remaining()));
+        try {
+            cipher.doFinal(cipherText, output);
+            output.flip();
+            return output;
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failure during decrypt", e);
+        }
+    }
+
+    @Override
+    public byte[] decrypt(byte[] cipherText)
+    {
+        return unwrapOrCopyToByteArray(decrypt(ByteBuffer.wrap(cipherText)));
+    }
+
+    @Override
+    public ByteBuffer encrypt(ByteBuffer plainText)
+    {
+        Cipher cipher = createEncryptCipher();
+        byte[] iv = cipher.getIV();
+        ByteBuffer output = ByteBuffer
+                .allocate(iv.length + cipher.getOutputSize(plainText.remaining()))
+                .put(iv);
+        try {
+            cipher.doFinal(plainText, output);
+            output.flip();
+            return output;
+        }
+        catch (GeneralSecurityException e) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, "Failure during encrypt", e);
+        }
+    }
+
+    @Override
+    public byte[] encrypt(byte[] plainText)
+    {
+        return unwrapOrCopyToByteArray(encrypt(ByteBuffer.wrap(plainText)));
+    }
+
+    @Override
+    public void destroy()
+    {
+        this.key = null;
+    }
+
+    @Override
+    public boolean isDestroyed()
+    {
+        SecretKeySpec key = this.key;
```

Comment:
- this variable is not needed: `return this.key == null || this.key.isDestroyed();`
- how can the key become destroyed? I don't see any calls to key.destroy().

Reply:
To address the first point, the `SecretKey` instance stored in a local variable before doing the checks against it because `AesSpillCipher.destroy()` sets `this.key = null;`. In the case of `AesSpillCipher.isDestroyed()` storing it in a temporary variable is just being paranoid- however in your later comment about `verifyKeyNotDestroyed()` the value of creating the local variable and returning it is to avoid a race condition between destruction and key usage that would raise an unhelpful `NullPointerException` if triggered instead of a `PrestoException` explaining the cipher was already destroyed.

On point 2... this is kind of a weird edge case in the JCE libraries. The `SecretKey` interface extends [Destroyable](https://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/cecd70d27b27/src/share/classes/javax/security/auth/Destroyable.java), but neither the `SecretKey` interface nor `SecretKeySpec` class overrides the `destroy()` method as of JDK 8, so the interface default implementation is used which unconditionally throws `DestroyFailedException`.

That left me with the option of either calling `key.destroy()` inside of `AesSpillCipher.destroy()` but catching and ignoring the `DestroyFailedException` (maybe someday the method will be actually implemented, right?) or not calling it at all. I chose to just not call the method, but left the `.isDestroyed()` check to be slightly defensive against future refactors. I'm fine with removing the check to `key.isDestroyed()` or even reintroducing the call to `key.destroy()` and ignoring the exception. Let me know what you think is best.

Reply:
@pettyjamesm James, thanks for the detailed explanation. I'm wondering now why do you need `destroy` method at all? Why not just let GC claim the object eventually? If there is a good reason to close the instance and prevent its usage after that, then it might be simpler to add `private boolean closed;` member variable and add `checkState(!closed, "...")` everywhere the key is used.

Reply:
For our use-cases, "destroying" the key after it's no longer necessary to use is desirable. It's an insurance policy against keys living longer than intended or being accidentally reused outside of the intended "one-per-spill-file" lifespan. By assigning the key to `null` in `AesSpillCipher#destroy()` we're both accomplishing the same functionality as `checkState(!closed, ...)` and making sure the secret key is eligible for reclamation by the GC even if the `AesSpillCipher` remains reachable for some reason.

I think leaving the check in for `key.isDestroyed()`, while paranoid- is probably preferable. I can see a world where users that enable spill encryption may want to manage key lifecycles differently than one-per-spill-file which might introduce the possibility of key originating and being destroyed outside of the `AesSpillCipher` at which point failing with a descriptive message is preferable to whatever exception would bubble from attempting to use a destroyed key. If you agree, I'm inclined to leave in the `key == null || key.isDestroyed()` check.